py/client/pydeephaven/agg.py

```diff
@@ -300,17 +313,52 @@ def std(cols: Union[str, List[str]] = None) -> Aggregation:
     return _AggregationColumns(agg_spec=agg_spec, cols=to_list(cols))
 
 
-def unique(cols: Union[str, List[str]] = None) -> Aggregation:
-    """Creates a Unique aggregation.
+def unique(cols: Union[str, List[str]] = None, include_nulls: bool = False,
+           non_unique_sentinel: Any = None) -> Aggregation:
+    """Creates a Unique aggregation which computes the single unique value within an aggregation group for each of
+    the given columns. If all values in a column are null, or if there is more than one distinct value in a column, the
+    result is null or the specified non_unique_sentinel value.
 
     Args:
         cols (Union[str, List[str]]): the column(s) to aggregate on, can be renaming expressions, i.e. "new_col = col";
             default is None, only valid when used in Table agg_all_by operation
+        include_nulls (bool): whether null is treated as a value for the purpose of determining if the values in the
+            aggregation group are unique, default is False.
+        non_unique_sentinel (Any): the non-null sentinel value when no unique value exists, default is None. Must be
+            a non-None value when include_nulls is True.
 
     Returns:
         an aggregation
     """
-    agg_spec = _GrpcAggSpec(unique=_GrpcAggSpec.AggSpecUnique(include_nulls=False, non_unique_sentinel=None))
+    if non_unique_sentinel is not None:
+        if isinstance(non_unique_sentinel, float):
+            if np.finfo(np.float32).min <= non_unique_sentinel <= np.finfo(np.float32).max:
+                agg_spec_non_unique_sentinel = _GrpcAggSpec.AggSpecNonUniqueSentinel(float_value=non_unique_sentinel)
+            else:
+                agg_spec_non_unique_sentinel = _GrpcAggSpec.AggSpecNonUniqueSentinel(double_value=non_unique_sentinel)
+        elif isinstance(non_unique_sentinel, int):
+            if np.iinfo(np.int8).min <= non_unique_sentinel <= np.iinfo(np.int8).max:
+                agg_spec_non_unique_sentinel = _GrpcAggSpec.AggSpecNonUniqueSentinel(byte_value=non_unique_sentinel)
+            elif np.iinfo(np.int16).min <= non_unique_sentinel <= np.iinfo(np.int16).max:
+                agg_spec_non_unique_sentinel = _GrpcAggSpec.AggSpecNonUniqueSentinel(short_value=non_unique_sentinel)
+            elif np.iinfo(np.int32).min <= non_unique_sentinel <= np.iinfo(np.int32).max:
+                agg_spec_non_unique_sentinel = _GrpcAggSpec.AggSpecNonUniqueSentinel(int_value=non_unique_sentinel)
+            else:
+                agg_spec_non_unique_sentinel = _GrpcAggSpec.AggSpecNonUniqueSentinel(long_value=non_unique_sentinel)
+        elif isinstance(non_unique_sentinel, str):
+            if len(non_unique_sentinel) == 1:
+                agg_spec_non_unique_sentinel = _GrpcAggSpec.AggSpecNonUniqueSentinel(
+                    char_value=ord(non_unique_sentinel))
+            else:
+                agg_spec_non_unique_sentinel = _GrpcAggSpec.AggSpecNonUniqueSentinel(string_value=non_unique_sentinel)
```

Comment:
I have a lot of concerns about this.  We are mapping to a type based on a value.  I'm always very concerned about code like this.

Reply:
@devinrsmith I was struggling with this, and I have to agree with @chipkent that it is ugly code. Without knowing the gRCP server implementation much, I am wondering if this type of mapping can be made easier for Python somehow?

Reply:
So, the python client can choose to go about this however it wants; at the RPC level though, you _will_ need some way to distinguish between all of these `<x>_value` cases though. Fundamentally, the java client does the same thing:

https://github.com/deephaven/deephaven-core/blob/v0.26.1/java-client/session/src/main/java/io/deephaven/client/impl/AggSpecBuilder.java#L206-L257

https://github.com/deephaven/deephaven-core/blob/v0.26.1/proto/proto-backplane-grpc/src/main/proto/deephaven/proto/table.proto#L868-L885

It uses a wrapper object though (as opposed to a plain `Object`/`object`): https://github.com/deephaven/deephaven-core/blob/v0.26.1/table-api/src/main/java/io/deephaven/api/object/UnionObject.java. The python client could do something similar:

```python
my_spec_1 = unique(..., non_unique_sentinal=int_(32))
my_spec_2 = unique(..., non_unique_sentinal=short_(32))
```